## **SCC5871/MAI5025 - INTRODUÇÃO A APRENDIZADO DE MÁQUINA**
## PROFA. ROSELI AP. FRANCELIN ROMERO
# 1°. SEMESTRE DE 2023
##

Alunos:
- Gabriel Bortoli (5378290)
- Lucas Maretti (6789752)



---
### Questão 01.

Para esse exercício, carregue a base de dados `dados.txt`. Esta base possui 14 variáveis independentes, organizados em 14 colunas. A 15ª coluna representa a variável dependente (o valor da função). Carregue os dados em uma estrutura Pandas DataFrame. Utilize um regressor linear nesse conjunto de dados e calcule seu erro quadrático médio (mean squared error - MSE)

- Avalie o MSE considerando o conjunto de dados **inteiro** para avaliação.

In [82]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score

data = pd.read_csv("dados.txt")

In [6]:
X = data.iloc[:,:-1]
Y = data.iloc[:,-1]

In [34]:
lin = LinearRegression()
lin.fit(X, Y)

Y_pred = lin.predict(X)

mse_case1 = mean_squared_error(Y,Y_pred)

print(f'MSE Calculado utilizando a Regressão Linear simples:\n{mse_case1:.4f}')


MSE Calculado utilizando a Regressão Linear simples:
3.8911


---
### Questão 02.

Agora faça regressão polinomial nos dados, utilizando um polinômio de grau 2. Calcule seu MSE.

* Dica: Utilize a classe importada para *transformar* os dados para uma regressão polinomial de grau 2. Mais informações sobre como o processo podem ser encontradas dentro da pasta do projeto.


In [33]:
poly2 = PolynomialFeatures(degree=2, include_bias=False)

X_poly2 = poly2.fit_transform(X)

lin.fit(X_poly2,Y)

Y_poly2 = lin.predict(X_poly2)

mse_case2 = mean_squared_error(Y,Y_poly2)

print(f'MSE Calculado utilizando a Regressão Polinomial de grau 2 e depois Regressão Linear:\n{mse_case2:.4f}')

MSE Calculado utilizando a Regressão Polinomial de grau 2 e depois Regressão Linear:
2.1378


---

### Questão 03.

Por fim, faça regressão polinomial utilizando um polinômio de grau 3. Verifique qual das 3 abordagens adotada produz o menor erro quadrático médio (MSE).
- Avalie o MSE considerando o conjunto de dados **inteiro**  para avaliação.


In [32]:
poly3 = PolynomialFeatures(degree=3, include_bias=False)

X_poly3 = poly3.fit_transform(X)

lin.fit(X_poly3,Y)

Y_poly3 = lin.predict(X_poly3)

mse_case3 = mean_squared_error(Y,Y_poly3)

print(f'MSE Calculado utilizando a Regressão Polinomial de grau 3 e depois Regressão Linear:\n{mse_case3:.4f}')

MSE Calculado utilizando a Regressão Polinomial de grau 3 e depois Regressão Linear:
0.0000


Podemos ver que a regressão polinomial de grau 3 conseguiu estabelecer parâmetros com exatidão para os dados propostos.

Os erros quadráticos diminuiram significativamente da regressão linear normal para grau 2 e depois zeraram para grau 3.

---

# Conjunto de dados de transações bancárias

No dataset que será utilizado para as próximas questões, cada elemento representa uma transação com cartões de crédito. 

A base possui 30 atributos. `Time` representa o tempo em segundos desde a primeira transação da base. `Amount` corresponde ao valor da transação.

Outros 28 atributos numéricos anonimizados também são fornecidos, com 1 variável (`Class`) de saída onde 1 significa fraude e 0 transação normal.

In [51]:
data_df = pd.read_csv("creditcard_sampled.csv", index_col=0)

In [52]:
# Removendo coluna com timestamps
data_df = data_df.drop(columns=["Time"])

print(data_df.head(3))
print(data_df.columns)

            V1        V2        V3        V4        V5        V6        V7  \
541  -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545 -2.537387   
623  -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823  0.325574   
4920 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788  0.562320   

            V8        V9       V10  ...       V21       V22       V23  \
541   1.391657 -2.770089 -2.772272  ...  0.517232 -0.035049 -0.465211   
623  -0.067794 -0.270953 -0.838587  ...  0.661696  0.435477  1.375966   
4920 -0.399147 -0.238253 -1.525412  ... -0.294166 -0.932391  0.172726   

           V24       V25       V26       V27       V28  Amount  Class  
541   0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623  -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920 -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  

[3 rows x 30 columns]
Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       '

### Questão 4

Divida o conjunto de dados em treino e teste seguindo as seguintes orientações:

- utilize a função [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) do sklearn
- utilize as seguintes proporções: 80% para treino e 20% para teste
- faça a divisão dos dados respeitando as proporções das classes (utilize o parâmetro *stratify*)

In [80]:
X_cc = data_df.iloc[:,:-1]
Y_cc = data_df.iloc[:,-1]

X_train, X_test, y_train, y_test = train_test_split(X_cc, Y_cc, test_size=0.2, random_state=42, stratify=Y_cc)

print(y_train.value_counts())
print(y_test.value_counts())

0    7999
1     394
Name: Class, dtype: int64
0    2001
1      98
Name: Class, dtype: int64


### Questão 5

Sabendo que estamos interessados em identificar fraudes e transações normais, responda:

- a) Identificação de fraudes é o problema de classificação ou regressão? Justifique sua resposta.
- b) Analise as [métricas do sklearn](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics), e escolha uma adequada para avaliar o problema de interesse. Justifique sua escolha.

a.) Considerando que os resultados esperados são se houve fraude ou não (1 e 0, respectivamente), estamos diante de um problema de classificação.

b.) Considerando que é importante minimizar falsos negativos (quando houve fraude e é considerado que não houve), uma boa métrica a ser usada é a de recall (TP/TP+FN).

---

### Questão 6

Considerando os conjuntos de dados dividos na Questão 4:

- Sabendo que os dados analisados correspondem a valores numéricos, aplique um procedimento adequado para pré-processamento dos mesmos (ex: normalização ou padronização) para os dados de treino e teste.
- Treine um [Regressor Logístico](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) nos dados de treino e avalie seu modelo no conjunto de teste **utilizando a métrica escolhida na Questão 5**.

In [84]:
# Pré processamento:

pipe_num = Pipeline(steps=[("scaler", MinMaxScaler())])

transform = ColumnTransformer(
    transformers=[
        ("transf num", pipe_num, X_cc.columns)
    ]
)

# Aplicando a transformação nos datasets:
X_train_new = pd.DataFrame(data = transform.fit_transform(X_train).round(4), columns= X_train.columns)
X_test_new = pd.DataFrame(data = transform.fit_transform(X_test).round(4), columns= X_test.columns)

####################

# Treinando o Regressor Logistico:

logreg = LogisticRegression(random_state=42).fit(X_train_new, y_train)
y_pred = logreg.predict(X_test_new)


# Avaliando o regressor:
print(f'Recall Score para a Regressão Logistica:\n{recall_score(y_test, y_pred):.4f}')

####################

# Avaliando outras métricas:

print(f'\n\nAccuracy Score para a Regressão Logistica:\n{accuracy_score(y_test, y_pred):.4f}')
print(f'F1 Score para a Regressão Logistica:\n{f1_score(y_test, y_pred):.4f}')
print(f'ROC AUC Score para a Regressão Logistica:\n{roc_auc_score(y_test, y_pred):.4f}')

Recall Score para a Regressão Logistica:
0.7653


Accuracy Score para a Regressão Logistica:
0.9890
F1 Score para a Regressão Logistica:
0.8671
ROC AUC Score para a Regressão Logistica:
0.8827


Podemos notar que para algumas métricas os resultados são muito bons, como para a acurácia, que ficou em 98.9%.

Porém, como é muito importante que todas as fraudes sejam pegas, ainda que algumas transações não fraudelentas também precisem ser validadas, é importante focar em melhorar o algoritmo para aumentar o Recall Score.

---